osu.Game.Tests/Visual/Gameplay/TestSceneStoryboardSamplePlayback.cs

```diff
@@ -47,26 +60,86 @@ public void TestStoryboardSamplesStopDuringPause()
             AddAssert("all storyboard samples stopped immediately", () => allStoryboardSamples.All(sound => !sound.IsPlaying));
 
             AddUntilStep("any storyboard samples playing after skip", () => allStoryboardSamples.Any(sound => sound.IsPlaying));
         }
 
-        private void checkForFirstSamplePlayback()
+        [TestCase(typeof(OsuModDoubleTime), 1.5)]
+        [TestCase(typeof(OsuModDoubleTime), 2)]
+        [TestCase(typeof(OsuModHalfTime), 0.75)]
+        [TestCase(typeof(OsuModHalfTime), 0.5)]
+        public void TestStoryboardSamplesPlaybackWithRateAdjustMods(Type expectedMod, double expectedRate)
+        {
+            AddStep("setup mod", () =>
+            {
+                ModRateAdjust testedMod = (ModRateAdjust)Activator.CreateInstance(expectedMod).AsNonNull();
+                testedMod.SpeedChange.Value = expectedRate;
+                storyboardMods = new[] { testedMod };
+            });
+
+            createPlayerTest(true);
+
+            AddAssert("sample playback rate matches mod rates", () => allStoryboardSamples.All(sound =>
+            {
+                return sound.ChildrenOfType<DrawableSample>().All(s => s.AggregateFrequency.Value == expectedRate);
+            }));
+        }
+
+        [TestCase(typeof(ModWindUp), 0.5, 2)]
+        [TestCase(typeof(ModWindUp), 1.51, 2)]
+        [TestCase(typeof(ModWindDown), 2, 0.5)]
+        [TestCase(typeof(ModWindDown), 0.99, 0.5)]
+        public void TestStoryboardSamplesPlaybackWithTimeRampMods(Type expectedMod, double initialRate, double finalRate)
+        {
+            AddStep("setup mod", () =>
+            {
+                ModTimeRamp testedMod = (ModTimeRamp)Activator.CreateInstance(expectedMod).AsNonNull();
+                testedMod.InitialRate.Value = initialRate;
+                testedMod.FinalRate.Value = finalRate;
+                storyboardMods = new[] { testedMod };
+            });
+
+            createPlayerTest(true);
+
+            ModTimeRamp gameplayMod = null;
+
+            AddUntilStep("mod speed change updated", () =>
+            {
+                gameplayMod = Player.GameplayState.Mods.OfType<ModTimeRamp>().Single();
+                return gameplayMod.SpeedChange.Value != initialRate;
+            });
+
+            AddAssert("sample playback rate matches mod rates", () => allStoryboardSamples.All(sound =>
+            {
+                return sound.ChildrenOfType<DrawableSample>().All(s => s.AggregateFrequency.Value == gameplayMod.SpeedChange.Value);
+            }));
+        }
+
+        private void createPlayerTest(bool skipIntro = false)
         {
+            CreateTest(null);
+
             AddAssert("storyboard loaded", () => Player.Beatmap.Value.Storyboard != null);
             AddUntilStep("any storyboard samples playing", () => allStoryboardSamples.Any(sound => sound.IsPlaying));
+
+            if (skipIntro)
+                AddStep("skip intro", () => InputManager.Key(Key.Space));
```

Comment:
I really don't like writing tests in this way. Can this be split to multiple helper methods? As in, `ensureStoryboardSamplesPlaying()` and `skipIntro()` or something, and then can each test call either just the first one or both depending on what they need?

Reply:
Fair call on the "ensure storyboard samples playing" part, which would also enforce separating the "skip intro" step to its method anyways.